invenio_rdm_records/services/access/service.py

```diff
@@ -0,0 +1,761 @@
+        # TODO deny if the identity already has access?
+        # TODO detect duplicate requests
+        record = self.record_cls.pid.resolve(id_)
+        self.require_permission(identity, "read", record=record)  # TODO
+
+        # TODO check that the user cannot actually access files
+        #      -> probably better with a new separate permission policy entry
+        denied = False
+        try:
+            self.require_permission(identity, "read_files", record=record)
+        except PermissionDeniedError:
+            denied = True
+
+        if not denied:
+            raise PermissionDeniedError()
+
+        # TODO refactor the duplicate detection logic
+        req_cls = current_requests_service.record_cls
+        model_cls = req_cls.model_cls
+        requests = [
+            request
+            for request in (
+                req_cls(rm.data, model=rm)
+                for rm in model_cls.query.filter(
+                    model_cls.json["created_by"] == {"user": str(identity.id)},
+                    model_cls.json["topic"] == {"record": id_},
+                )
+                if rm.data and rm.data["type"] == UserAccessRequest.type_id
+            )
+            if request.is_open
+        ]
+
+        if any((req.is_open for req in requests)):
+            raise DuplicateAccessRequestError([str(r.id) for r in requests])
+
+        record = self.record_cls.pid.resolve(id_)
+        data = {
+            "payload": {
+                "permission": "view",
+                "message": message,
+            }
+        }
+
+        # TODO might also be community managers
+        receiver = None
+        record_owner = record.parent.access.owner.resolve()
+        if record_owner:
+            receiver = record_owner
+
+        if receiver is None:
+            # TODO as fallback maybe admins?
+            pass
+
+        request = current_requests_service.create(
+            identity,
+            data,
+            UserAccessRequest,
+            receiver,
+            topic=record,
+            expires_at=None,  # TODO expiration date
+            expand=expand,
+            uow=uow,
+        )
+
+        # immediately submit the request, unless it has errors
+        # TODO is this correct?
+        if request.errors:
+            return request
+
+        message = {
+            "payload": {
+                "content": data["payload"].get("message") or _("No comment"),
+            }
+        }
+
+        # TODO can we pass the message here?
+        return current_requests_service.execute_action(
+            identity,
+            request.id,
+            "submit",
+            data=message,
+            uow=uow,
+        )
+
+    @unit_of_work()
+    def create_guest_access_request_token(
+        self, identity, id_, data, expand=False, uow=None
+    ):
+        """Create an request token that can be used to create an access request."""
+        # TODO proper permission checks
```

Comment:
The permissions already work as expected, but the code for that could use a bit of face-lifting: Rather than having these awkward code constructs in the `access_service.create_*` methods right now, we should have new generators and new entries in the permission policy.